src/rgw/rgw_dedup_utils.h

```diff
@@ -85,6 +88,93 @@ namespace rgw::dedup {
     uint8_t flags;
   };
 
+  class Throttle {
+  public:
+    // @max_calls_per_sec - max requests per second allowed, 0 means unlimited
+    // disbaled by default
+    Throttle(size_t max_calls_per_sec=0) {
+      reset(max_calls_per_sec);
+    }
+
+    inline size_t get_max_calls_per_second() const {
+      return max_calls;
+    }
+
+    inline uint64_t get_sleep_events() const {
+      return sleep_events;
+    }
+
+    inline uint64_t get_sleep_time_usec() const {
+      return sleep_time_usec;
+    }
+
+    inline bool is_disabled() const {
+      return disabled;
+    }
+
+    // set the number of calls per second
+    // zero means unlimited
+    void reset(size_t max_calls_per_sec) {
+      std::unique_lock lock(reset_mutex);
+      max_calls = max_calls_per_sec;
+      tokens = max_calls_per_sec;
+      last_reset.store(steady_clock::now());
+
+      if (max_calls_per_sec == 0) {
+        disabled = true;
+      }
+      else {
+        disabled = false;
+      }
+    }
+
+    // Blocks until allowed to proceed
+    void acquire() {
+      if (is_disabled()) {
+        return;
+      }
+      // get last_reset time atomically to prevent a race with ouside reset call
+      const steady_clock::time_point last_reset_copy = last_reset.load();
+      const steady_clock::time_point now = steady_clock::now();
+
+      uint64_t elapsed_usec = duration_cast<microseconds>(now - last_reset_copy).count();
```

Comment:
won't you have a race condition here?
* one worker loaded last _reset that was less than a second ago (e.g. 900ms passed)
* another worker loaded and decided that more than a second had passed (1100ms passed). and called reset
* the first worker will calculate a very short sleep time. in the above case, 100ms, and then reset